### ISSUE TITLE
With reportUnknownMemberType, generic __init__ wants specialization which causes errors at runtime

### ISSUE THREAD

**Issue by Wizzerinus** [microsoft/pyright#10296]
**Describe the bug**
If `GenericClass.__init__()` is called, Pyright complains that the GenericClass is not specialized. Trying to specialize it `GenericClass[int]` suppresses the error, but causes an error at runtime:

```
  File "/home/wizzerinus/tests/generic_erasure.py", line 19, in <module>
    C()
    ~^^
  File "/home/wizzerinus/tests/generic_erasure.py", line 15, in __init__
    A[int].__init__(self)
    ~~~~~~~~~~~~~~~^^^^^^
TypeError: _GenericAlias.__init__() missing 1 required positional argument: 'args'
```

Playground reproduction:

https://pyright-play.net/?strict=true&code=MYGwhgzhAEBCDaAVAugLgLACho%2BgNzBAFcBTVaRLLUSGAYQAoEBLAOwBdkBKDbXAExIAzaAH1RbZu3EMIJEEJ5ZcKuADpxk6aNnzFVTNXBRoAESbw2nJXxyCRm1lJlyFN1bhYdkGiU%2B26blhAA

**Code or Screenshots**

```python
# pyright: reportUnknownMemberType=error

class A[T]:
    pass

class B(A[int]):
    def __init__(self):
        A.__init__(self)  # pyright ERR

class C(A[int]):
    def __init__(self):
        A[int].__init__(self)  # pyright OK

B()   # runtime OK
C()   # runtime ERR
```

**VS Code extension or command-line**
CLI, pyright 1.1.398

**Comment by erictraut** [comment]
Pyright is working as intended here. The expression `A.__init__` is specialized to `A[Unknown].__init__` because you haven't provided a type argument for `A`, and the typing spec says that type checkers should assume `Any` in this case. Pyright uses `Unknown` to mean an "implied `Any`".

As for the runtime error that you're seeing in `C`, I can't explain that. This code looks completely legit to me, and other type checkers like mypy don't flag it as an error. My best guess is that it's a bug in the implementation of `_GenericAlias` — or at least an unexpected limitation, but it's hard to say because this is an undocumented internal type. For most attributes, `_GenericAlias` forwards accesses to the original type, but maybe it excepts `__init__` because it's needed for the construction of the `_GenericAlias` object. Maybe @JelleZijlstra could provide additional insight.

In any case, I don't think this is a bug in pyright. It's working as intended, and its behavior is consistent with other type checkers here.

**Comment by Wizzerinus** [comment]
Well, that sucks. I used to imagine typecheckers existed to prevent bugs, not to create them. Silly me, I guess.

**Comment by erictraut** [comment]
Type checkers are designed to prevent bugs in common coding patterns, but they need to work within a framework of standardized and consistent rules. Without such rules, the type system — and type checking behaviors — would be nearly impossible to reason about or maintain. Type checkers don't claim to catch all possible bugs. There are many classes of bugs that they cannot catch for a variety of reasons including odd and inconsistent behaviors in the Python runtime.

There are cases where type checker maintainers and/or the typing community as a whole decide to break these rules to accommodate a common usage pattern or a common source of bugs. We try to keep such exceptions to a minimum, however. I don't think this case qualifies as a common pattern. To the best of my recollection, you're the first person to report this issue in either the pyright or mypy issue trackers.

**Comment by JelleZijlstra** [comment]
You probably should be using `super().__init__()`.

**Comment by erictraut** [comment]
Yeah, good point. Thanks Jelle.

**Comment by Wizzerinus** [comment]
I cannot use super().__init__() as I have a multiple inheritance class and I need to construct the second superclass in the MRO. Oh well.
